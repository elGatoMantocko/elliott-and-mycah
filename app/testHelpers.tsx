import { render, RenderOptions, RenderResult } from '@testing-library/react';
import { match } from 'css-mediaquery';
import React from 'react';
import { MemoryRouter, Route, Routes, useLocation } from 'react-router';

type TestLocations = 'route-location';

/**
 * The element testid used to show the location.
 */
export const TestRouteLocation: TestLocations = 'route-location';

interface TestLocationProps {
  testId?: string;
}

/**
 * Component that exposes the location path and search to the dom
 * @param root0 props
 * @param root0.testId test id passed to the inner location element
 * @returns element
 */
const TestLocation = ({ testId }: TestLocationProps) => {
  const location = useLocation();
  return (
    <div data-testid={testId ?? TestRouteLocation}>{`${location.pathname}${location.search}`}</div>
  );
};

interface RouterRenderOptions extends RenderOptions {
  locationTestId?: string;
}

/**
 * Use to render UI with a react-router with the ability to search for a `route-location` test id.
 * @example
 * ```typescriptreact
 * const el = renderWithRouter(<MyComponent />);
 * const locationElementData = await el.findByTestId(TestRouteLocation);
 * ```
 * @param ui to render in the test
 * @param options passed to testing-library
 * @param options.locationTestId passed to the `<TestLocation />` element
 * @returns render result
 */
export const renderWithRouter = (
  ui: React.ReactElement,
  { locationTestId, ...renderOptions }: RouterRenderOptions = {},
): RenderResult => {
  return render(
    <MemoryRouter>
      {ui}
      <Routes>
        <Route path="*" element={<TestLocation testId={locationTestId} />} />
      </Routes>
    </MemoryRouter>,
    renderOptions,
  );
};

// eslint-disable-next-line @typescript-eslint/no-empty-function
const noop = () => {};

/**
 * Function to create a match media for mocking media queries.
 * @param width to mock in the jsdom
 * @returns matchMedia options
 */
export function createMatchMedia(width: number): (query: string) => MediaQueryList {
  return (query: string) =>
    ({
      matches: match(query, { width }),
      addListener: noop,
      removeListener: noop,
    }) as unknown as MediaQueryList;
}
