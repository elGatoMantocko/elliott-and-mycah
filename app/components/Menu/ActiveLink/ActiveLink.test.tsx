import { act } from '@testing-library/react';
import React from 'react';

import { renderWithRouter, TestRouteLocation } from '../../../testHelpers';
import { ActiveLink } from '.';

it('should render a link', async () => {
  const el = renderWithRouter(
    <ActiveLink data-testid="basic-link" href="/test">
      Some test link
    </ActiveLink>,
  );

  expect(await el.findByTestId('basic-link')).toMatchSnapshot('basic-link');
});

it('should render disabled link', async () => {
  const el = renderWithRouter(
    <ActiveLink data-testid="disabled-link" disabled href="/test">
      Some test link
    </ActiveLink>,
  );

  expect(await el.findByTestId('disabled-link')).toMatchSnapshot('disabled-link');

  await act(async () => {
    const link = await el.findByTestId('disabled-link');
    link.click();
  });

  // nothing should change after clicking a disabled link
  expect(el.container).toMatchSnapshot('disabled-link');
});

it('should render active on the current path', async () => {
  const el = renderWithRouter(
    <ActiveLink data-testid="some-test-link" href="/">
      Some test link
    </ActiveLink>,
  );

  expect(await el.findByTestId('some-test-link')).toMatchSnapshot('some-test-link');
});

it('should render a link with null href', async () => {
  const el = renderWithRouter(<ActiveLink data-testid="null-href">Some test link</ActiveLink>);

  await act(async () => {
    const link = await el.findByTestId('null-href');
    link.click();
  });

  const locationEl = await el.findByTestId(TestRouteLocation);
  expect(locationEl.textContent).toEqual('/');

  expect(await el.findByTestId('null-href')).toMatchSnapshot('null-href');
});

it('should navigate when a link is clicked', async () => {
  const el = renderWithRouter(
    <ActiveLink data-testid="with-href" href="/test-path">
      Some test link
    </ActiveLink>,
  );

  await act(async () => {
    const link = await el.findByTestId('with-href');
    link.click();
  });

  const locationEl = await el.findByTestId(TestRouteLocation);
  expect(locationEl.textContent).toEqual('/test-path');

  expect(await el.findByTestId('with-href')).toMatchSnapshot('with-href');
});
