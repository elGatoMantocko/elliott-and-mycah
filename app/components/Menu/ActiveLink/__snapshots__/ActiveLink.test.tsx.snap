// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should navigate when a link is clicked: with-href 1`] = `
<a
  class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-ur4oxz-MuiTypography-root-MuiLink-root"
  data-testid="with-href"
>
  Some test link
</a>
`;

exports[`should render a link with null href: null-href 1`] = `
<a
  class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-qztcla-MuiTypography-root-MuiLink-root"
  data-testid="null-href"
>
  Some test link
</a>
`;

exports[`should render a link: basic-link 1`] = `
<a
  class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-qztcla-MuiTypography-root-MuiLink-root"
  data-testid="basic-link"
>
  Some test link
</a>
`;

exports[`should render active on the current path: some-test-link 1`] = `
<a
  class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-ur4oxz-MuiTypography-root-MuiLink-root"
  data-testid="some-test-link"
>
  Some test link
</a>
`;

exports[`should render disabled link: disabled-link 1`] = `
<a
  class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineNone css-1tgc2fw-MuiTypography-root-MuiLink-root"
  data-testid="disabled-link"
>
  Some test link
</a>
`;

exports[`should render disabled link: disabled-link 2`] = `
<div>
  <a
    class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineNone css-1tgc2fw-MuiTypography-root-MuiLink-root"
    data-testid="disabled-link"
  >
    Some test link
  </a>
  <div
    data-testid="route-location"
  >
    /
  </div>
</div>
`;
